File: Notificare.Assets.Core/Models/NotificareAsset.cs
namespace NotificareSdk.Assets.Core.Models;

public class NotificareAsset
{
    public string Title { get; }
    public string? Description { get; }
    public string? Key { get; }
    public string? Url { get; }
    public NotificareAssetButton? Button { get; }
    public NotificareAssetMetaData? MetaData { get; }
    public IDictionary<string, object> Extra { get; }

    public NotificareAsset(string title, string? description, string? key, string? url, NotificareAssetButton? button,
        NotificareAssetMetaData? metaData, IDictionary<string, object> extra)
    {
        Title = title;
        Description = description;
        Key = key;
        Url = url;
        Button = button;
        MetaData = metaData;
        Extra = extra;
    }
}

public class NotificareAssetButton
{
    public string? Label { get; }
    public string? Action { get; }

    public NotificareAssetButton(string? label, string? action)
    {
        Label = label;
        Action = action;
    }
}

public class NotificareAssetMetaData
{
    public string OriginalFileName { get; }
    public string ContentType { get; }
    public int ContentLength { get; }

    public NotificareAssetMetaData(string originalFileName, string contentType, int contentLength)
    {
        OriginalFileName = originalFileName;
        ContentType = contentType;
        ContentLength = contentLength;
    }
}

File: Notificare.Assets.iOS.Binding.Bridge/Sources/Models/NotificareAsset.swift
import Foundation
import NotificareAssetsKit

@objc
public class NotificareAsset : NSObject {
    @objc public let title: String
    @objc public let assetDescription: String?
    @objc public let key: String?
    @objc public let url: String?
    @objc public let button: NotificareAssetButton?
    @objc public let metaData: NotificareAssetMetaData?
    @objc public let extra: [String: Any]

    @objc public init(title: String, assetDescription: String?, key: String?, url: String?, button: NotificareAssetButton?, metaData: NotificareAssetMetaData?, extra: [String : Any]) {
        self.title = title
        self.assetDescription = assetDescription
        self.key = key
        self.url = url
        self.button = button
        self.metaData = metaData
        self.extra = extra
    }

    public convenience init(from asset: NotificareAssetsKit.NotificareAsset) {
        self.init(
            title: asset.title,
            assetDescription: asset.description,
            key: asset.key,
            url: asset.url,
            button: asset.button.map { NotificareAssetButton(from: $0) },
            metaData: asset.metaData.map { NotificareAssetMetaData(from: $0) },
            extra: asset.extra
        )
    }
}

@objc
public class NotificareAssetButton : NSObject {
    @objc public let label: String?
    @objc public let action: String?

    @objc public init(label: String?, action: String?) {
        self.label = label
        self.action = action
    }

    public convenience init(from button: NotificareAssetsKit.NotificareAsset.Button) {
        self.init(
            label: button.label,
            action: button.action
        )
    }
}

@objc
public class NotificareAssetMetaData : NSObject {
    @objc public let originalFileName: String
    @objc public let contentType: String
    @objc public let contentLength: Int

    @objc public init(originalFileName: String, contentType: String, contentLength: Int) {
        self.originalFileName = originalFileName
        self.contentType = contentType
        self.contentLength = contentLength
    }

    public convenience init(from metaData: NotificareAssetsKit.NotificareAsset.MetaData) {
        self.init(
            originalFileName: metaData.originalFileName,
            contentType: metaData.contentType,
            contentLength: metaData.contentLength
        )
    }
}
